docs/user_guide/multi-stage.md

```diff
@@ -0,0 +1,94 @@
+```
+
+The following table summarizes the behavior of `zero_stage`:
+
+| `zero_stage`  | Weights    | Gradients  | Optimizer States | Communication overhead    |
+| ------------- | ---------- | ---------- | ---------------- | ------------------------- |
+| `1` (default) | Replicated | Replicated | Sharded          | Moderate, default choice  |
+| `2`           | Replicated | Sharded    | Sharded          | Moderate[^1]              |
+| `3`           | Sharded    | Sharded    | Sharded          | High[^2]                  |
+
+[^1]: Communication overhead for ZeRO Stage 2 is similar to Stage 1, except during (depth-first) gradient accumulation when additional all-reduce operations occur.
+[^2]: Communication overhead for ZeRO Stage 3 is higher than Stage 2, especially during (depth-first) gradient accumulation.
+
+Optimizer states are always sharded by default. ZeRO Stage 0 (full replication) is not supported.
+
+While ZeRO Stage 3 introduces the most communication overhead, the practical difference between Stages 1 and 2 is minimal except during gradient accumulation.
+
+**Recommendation:**
+
+- **ZeRO Stage 1 (default)**: Ideal for most training scenarios.
+- **ZeRO Stage 2**: Useful if gradients cause memory pressure.
+- **ZeRO Stage 3**: Useful for very large models exceeding GPU memory.
+
+In general, start with the default (`zero_stage: 1`) and verify if your model trains without memory errors. If you encounter out-of-memory issues, try increasing `zero_stage`:
+
+```yaml
+multi_stage:
+  zero_stage: 2
+```
+
+Increasing ZeRO-style sharding reduces memory consumption but may add communication overhead between GPUs or nodes, potentially slowing down training. Before increasing `zero_stage`, first try lowering the micro batch size or sequence length, as this typically incurs less overhead.
+
+You'll likely iterate between adjusting `zero_stage`, micro batch size, and sequence length to find the optimal balance of memory usage and training throughput. If these adjustments don't resolve your issue, or you're unsatisfied with tradeoffs like sequence length versus throughput, reconsider your broader parallelism strategy. This includes adjusting tensor parallelism, pipeline parallelism, or sequence data parallelism, covered in greater depth in the [Parallelism Guide](parallelism.md).
+
+## Expert Options
+
+Beyond `zero_stage`, Fast-LLM offers additional multi-stage settings for fine-tuning. These advanced options typically don't need manual adjustment. Change them only if you're certain about your goals and tradeoffs.
+
+### Buffers
+
+When gradients or weights are sharded, Fast-LLM accumulates partial results in shared *buffers* during forward and backward passes, separately for gradients and weights. These buffers reduce communication overhead by batching gradient or weight updates across GPUs or nodes. The options `num_grad_buffers` and `num_weight_buffers` control the number of buffers used for gradients and weights, respectively.
```

Comment:
Might be useful to state explicitly how this relates to ZeRO stages:
* `num_layers` buffers: Store all layers in memory, as in traditional (non-ZeRO) DP
* `2` Keep weights/gradients one layer at the time, i.e. ZeRO stage 2/3. Second buffer is there for network overlap.